pkg/utils/volumeattachment/volumeattachment.go

```diff
@@ -0,0 +1,80 @@
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package volumeattachment
+
+import (
+	"context"
+
+	"github.com/samber/lo"
+
+	"sigs.k8s.io/karpenter/pkg/utils/pod"
+
+	v1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	nodeutil "sigs.k8s.io/karpenter/pkg/utils/node"
+	volumeutil "sigs.k8s.io/karpenter/pkg/utils/volume"
+)
+
+// FilterVolumeAttachments filters out volumeAttachments that should not block the termination of the passed node
+func FilterVolumeAttachments(ctx context.Context, kubeClient client.Client, node *v1.Node, volumeAttachments []*storagev1.VolumeAttachment) ([]*storagev1.VolumeAttachment, error) {
+	var filteredVolumeAttachments []*storagev1.VolumeAttachment
+	// No need to filter empty volumeAttachments list
+	if len(volumeAttachments) == 0 {
+		return volumeAttachments, nil
+	}
+	// Filter out non-drain-able pods
+	pods, err := nodeutil.GetPods(ctx, kubeClient, node)
+	if err != nil {
+		return nil, err
+	}
+	drainablePods := lo.Reject(pods, func(p *v1.Pod, _ int) bool {
+		return pod.ToleratesDisruptionNoScheduleTaint(p)
```

Comment:
I think we also need to exclude [static pods](https://kubernetes.io/docs/tasks/configure-pod-container/static-pod/) here. I'm also thinking we should exclude volume attachments from pods which are [stuck terminating](https://github.com/kubernetes-sigs/karpenter/blob/a122e25df3decf77daf0930d5fb11f7c91c6b7ca/pkg/utils/pod/scheduling.go#L132-L138). If a pod gets stuck terminating past it's TGP, we want to be able to proceed and terminate the EC2 instance.

Reply:
> I think we also need to exclude static pods here.

I'm not sure if static pods can even have volumeattachments because "The spec of a static Pod cannot refer to other API objects," but will double check manually. 

> If a pod gets stuck terminating past it's TGP, we want to be able to proceed and terminate the EC2 instance.

Does the current PR address this implicitly by skipping waitForVolumeDetachments if after TGP? 

I believe the 1 second reconciler requeue if we hit TGP time during waitForVolumeDetachments is a good feature (1 sec is palatable delay for customers, but 1 sec is enough time for a few more volumes to get unmounted such that we avoid 6+ min delay)

Reply:
Yeah static pods are not allowed to have non host-path and empty dir volumes. Therefore no volume attachment objects. 

"K/K #103587: Kubelet should enforce that static pods do not reference other API objects" mentions that the mirrored pod will fail to create because of API validation, even if kubelet doesn't enforce it.

Reply:
Sounds good, thanks for looking into the static pod case. 

> Does the current PR address this implicitly by skipping waitForVolumeDetachments if after TGP?

Not necessarily, since TGP isn't a required field. If a user doesn't configure TGP, they could potentially be stuck here indefinitely, whereas before Karpenter would proceed to terminate the instance.